### ISSUE TITLE
codec = flac in conjunction with MPD / pipe input causes stuttering

### ISSUE THREAD

**Issue by nickalcock** [badaix/snapcast#942]
Observed with snapcast 0.25.0 on both a Raspberry Pi 3 and a homebrewed Linux desktop box.

The rpi 3 is using the alsa backend and is installed from distributed nopulse packages; an mpd and a kodi are already running on the same machine, but they're not playing anything right now, and since no software nor hardware mixer exists on the rpi 3, we can be sure that when playback is happening, snapclient has exclusive control of the sound device, which is an HDMI output. (It doesn't help.) It's connected via a not terribly fast wifi link, but one which is more than good enough to view H.264-compressed 1080p video: a bunch of flac frames should be fine (and are fine, when playing flacs back using mpd).

The Linux desktop box is using the pulseaudio backend and is running snapcast 0.25.0 built from source: it's connected directly to the snapserver box via an uncontended 10GbE link. Both PA and snapclient are running as the logged-in user (because PulseAudio likes it that way). The snapserver is also running 0.25.0. We can completely rule out the network as a cause, I think: 10GbE is *definitely* enough.

The audio is coming from an mpd pipe output configured as in the documentation.

Active server-side config (all else left at defaults):

source = pipe:///tmp/snapfifo?name=Housewide
codec = flac
send_to_muted = false

The stuttering is severe enough, well over ten stutters a second, that it turns a lot of things (anything with vocals, piano music) into something like the audio equivalent of moire patterns: nothing like the input at all. It's quite possible that we're getting one stutter per 20ms (i.e. one stutter per chunk).

Boosting the buffer size to 2000ms and the chunk size to 100ms makes it *disastrously* worse, with 100ms stutters of silence, and latency so high that it takes about a minute for the sound to stop on the client after I hit pause on the server. Experimentation shows that the culprit is the chunk size: raising it above the default causes disastrous stuttering, but unfortunately lowering it doesn't help fix the stuttering at all. :(

Logs:

nix@silk 130 /home/nix% snapclient --logfilter '*:debug' --player pulse
2021-11-20 15-37-32.220 [Debug] (Snapclient) Trying to get PCM device for player: pulse, parameter: , card: default
2021-11-20 15-37-32.242 [Info] (Snapclient) Version 0.25.0
2021-11-20 15-37-32.255 [Info] (Avahi) (Browser) NEW: service 'Snapcast' of type '_snapcast._tcp' in domain 'local'
2021-11-20 15-37-32.255 [Info] (Avahi) (Browser) CACHE_EXHAUSTED
2021-11-20 15-37-32.414 [Info] (Avahi) Service 'Snapcast' of type '_snapcast._tcp' in domain 'local':
2021-11-20 15-37-32.414 [Info] (Avahi) 	loom.local:1704 (192.168.16.8)
2021-11-20 15-37-32.414 [Debug] (Avahi) 	TXT=
2021-11-20 15-37-32.414 [Debug] (Avahi) 	Proto=0
2021-11-20 15-37-32.414 [Debug] (Avahi) 	cookie is 0
2021-11-20 15-37-32.414 [Debug] (Avahi) 	is_local: 0
2021-11-20 15-37-32.414 [Debug] (Avahi) 	our_own: 0
2021-11-20 15-37-32.414 [Debug] (Avahi) 	wide_area: 0
2021-11-20 15-37-32.414 [Debug] (Avahi) 	multicast: 1
2021-11-20 15-37-32.414 [Debug] (Avahi) 	cached: 1
2021-11-20 15-37-32.414 [Info] (Controller) Found server 192.168.16.8:1704
2021-11-20 15-37-32.414 [Info] (Connection) Resolving host IP for: 192.168.16.8
2021-11-20 15-37-32.414 [Info] (Connection) Connecting
2021-11-20 15-37-32.415 [Notice] (Connection) Connected to 192.168.16.8
2021-11-20 15-37-32.415 [Info] (Connection) My MAC: "00:1b:21:e3:a4:d6", socket: 8
2021-11-20 15-37-32.425 [Debug] (Connection) outstanding async_write
2021-11-20 15-37-32.426 [Info] (Controller) ServerSettings - buffer: 1000, latency: 0, volume: 100, muted: 0
metadata:{"STREAM":"Housewide"}
2021-11-20 15-37-32.426 [Info] (Controller) Codec: flac, sampleformat: 48000:16:2
2021-11-20 15-37-32.426 [Info] (Player) Player name: pulse, device: default, description: Let PulseAudio server choose the device, idx: 1, sharing mode: unspecified, parameters: <none>
2021-11-20 15-37-32.426 [Info] (Player) Mixer mode: software, parameters: <none>
2021-11-20 15-37-32.426 [Info] (Player) Sampleformat: 48000:16:2, stream: 48000:16:2
2021-11-20 15-37-32.426 [Info] (PulsePlayer) Setting property "application.icon_name" to "snapcast"
2021-11-20 15-37-32.426 [Info] (PulsePlayer) Setting property "media.role" to "music"
2021-11-20 15-37-32.426 [Info] (PulsePlayer) Using buffer_time: 100 ms, server: default
2021-11-20 15-37-32.427 [Debug] (PulsePlayer) State changed 2: authorizing
2021-11-20 15-37-32.433 [Debug] (PulsePlayer) State changed 3: setting name
2021-11-20 15-37-32.446 [Debug] (PulsePlayer) State changed 4: ready
2021-11-20 15-37-32.447 [Debug] (Player) setVolume exp with base 10: 1 => 1
2021-11-20 15-37-32.447 [Info] (Stream) outputBufferDacTime > bufferMs: 25227 > 1000
2021-11-20 15-37-32.447 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-37-32.459 [Info] (Controller) diff to server [ms]: 1.11652e+09
2021-11-20 15-37-32.467 [Info] (Stream) No chunks available
2021-11-20 15-37-33.018 [Info] (Stream) No chunks available
2021-11-20 15-37-33.450 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-37-34.004 [Info] (Stream) No chunks available
2021-11-20 15-37-34.456 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-37-35.008 [Info] (Stream) No chunks available
2021-11-20 15-37-35.457 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-37-36.011 [Info] (Stream) No chunks available
2021-11-20 15-37-36.461 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-37-37.015 [Info] (Stream) No chunks available
2021-11-20 15-37-37.463 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-37-38.010 [Info] (Stream) No chunks available
2021-11-20 15-37-38.477 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-37-39.006 [Info] (Stream) No chunks available
2021-11-20 15-37-40.312 [Debug] (Stream) Silent frames: 265, frames: 969, age: -5.524
2021-11-20 15-37-40.333 [Debug] (Stats) Chunk: 0	0	0	0	1	51	0
2021-11-20 15-37-41.004 [Debug] (Stats) Chunk: 0	0	0	0	34	52	0
2021-11-20 15-37-42.006 [Debug] (Stats) Chunk: 0	0	0	0	83	52	0
2021-11-20 15-37-43.010 [Debug] (Stats) Chunk: 0	0	0	0	132	51	0
2021-11-20 15-37-44.013 [Debug] (Stats) Chunk: 0	0	0	0	181	51	0
2021-11-20 15-37-45.015 [Debug] (Stats) Chunk: 0	0	0	0	230	52	0
2021-11-20 15-37-46.002 [Debug] (Stats) Chunk: -1	0	0	0	278	51	0
2021-11-20 15-37-47.008 [Debug] (Stats) Chunk: -1	-1	-1	0	327	52	0
2021-11-20 15-37-48.011 [Debug] (Stats) Chunk: 0	0	-1	0	376	52	2
2021-11-20 15-37-49.014 [Debug] (Stats) Chunk: 0	0	0	0	425	52	1
2021-11-20 15-37-50.018 [Debug] (Stats) Chunk: 0	0	0	0	474	51	0
2021-11-20 15-37-51.018 [Debug] (Stats) Chunk: 0	0	0	0	500	52	0
2021-11-20 15-37-52.020 [Debug] (Stats) Chunk: 0	0	0	0	500	52	0
2021-11-20 15-37-53.001 [Debug] (Stats) Chunk: 0	0	0	0	500	52	0
2021-11-20 15-37-54.015 [Debug] (Stats) Chunk: -1	0	0	0	500	52	0
2021-11-20 15-37-55.012 [Debug] (Stats) Chunk: -1	-1	-1	0	500	52	0
2021-11-20 15-37-56.010 [Debug] (Stats) Chunk: -1	0	-1	0	500	52	0
2021-11-20 15-37-57.004 [Debug] (Stats) Chunk: -1	-1	-1	0	500	52	2
2021-11-20 15-37-58.005 [Debug] (Stats) Chunk: 0	0	-1	0	500	51	2
2021-11-20 15-37-58.208 [Info] (Stream) Exception: Not enough frames available, requested frames: 986, available: 631
2021-11-20 15-37-58.208 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-37-58.229 [Debug] (Stream) age > 0: 7ms, dropping old chunks
2021-11-20 15-37-59.007 [Info] (Stream) No chunks available
2021-11-20 15-37-59.210 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-38-00.006 [Info] (Stream) No chunks available
2021-11-20 15-38-00.210 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-38-01.010 [Info] (Stream) No chunks available
2021-11-20 15-38-01.214 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-38-02.017 [Info] (Stream) No chunks available
2021-11-20 15-38-02.219 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-38-03.018 [Info] (Stream) No chunks available
2021-11-20 15-38-03.222 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-38-04.004 [Info] (Stream) No chunks available
2021-11-20 15-38-04.228 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-38-05.007 [Info] (Stream) No chunks available
2021-11-20 15-38-05.230 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-38-06.009 [Info] (Stream) No chunks available
2021-11-20 15-38-06.235 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 15-38-07.017 [Info] (Stream) No chunks available
2021-11-20 15-38-07.242 [Debug] (Stream) Failed to get chunk, returning silence
^C2021-11-20 15-38-07.354 [Info] (Snapclient) Received signal 2: Interrupt
2021-11-20 15-38-07.354 [Debug] (PulsePlayer) Destructor
2021-11-20 15-38-07.354 [Info] (PulsePlayer) Stop
2021-11-20 15-38-07.354 [Debug] (PulsePlayer) State changed 6: terminated
2021-11-20 15-38-07.355 [Debug] (Connection) Disconnecting
2021-11-20 15-38-07.355 [Debug] (Connection) Disconnected
2021-11-20 15-38-07.355 [Notice] (Snapclient) Snapclient terminated.

The logs report a lot  of -1s, but the crackling is continuous even when no -1 appears.

**Comment by badaix** [comment]
Didn't read the long text yet, but the logs tell me that the client stops receiving data from the server. Please try with a wired connection if the problem persists. Or let a ping to the server  run in parallel to the client.

**Comment by nickalcock** [comment]
Very strange! This is with a 10GbE wired connection (and my NFSv4-mounted $HOME is run over that same connection, so if it stuttered even for a moment, I really would notice), but the logs cover the time from before playback started until the time after it finished. Running it for longer, starting playing at 18:56:30 for 30 seconds, and then halting playback before terminating snapclient, shows (with stuttering continuous and network connection uninterrupted):

```
nix@silk 137 /home/nix% snapclient --logfilter '*:debug' --player pulse
2021-11-20 18-56-15.667 [Debug] (Snapclient) Trying to get PCM device for player: pulse, parameter: , card: default
2021-11-20 18-56-15.690 [Info] (Snapclient) Version 0.25.0
2021-11-20 18-56-15.702 [Info] (Avahi) (Browser) NEW: service 'Snapcast' of type '_snapcast._tcp' in domain 'local'
2021-11-20 18-56-15.703 [Info] (Avahi) (Browser) CACHE_EXHAUSTED
2021-11-20 18-56-15.703 [Info] (Avahi) Service 'Snapcast' of type '_snapcast._tcp' in domain 'local':
2021-11-20 18-56-15.703 [Info] (Avahi) 	loom.local:1704 (192.168.16.8)
2021-11-20 18-56-15.703 [Debug] (Avahi) 	TXT=
2021-11-20 18-56-15.703 [Debug] (Avahi) 	Proto=0
2021-11-20 18-56-15.703 [Debug] (Avahi) 	cookie is 0
2021-11-20 18-56-15.703 [Debug] (Avahi) 	is_local: 0
2021-11-20 18-56-15.703 [Debug] (Avahi) 	our_own: 0
2021-11-20 18-56-15.703 [Debug] (Avahi) 	wide_area: 0
2021-11-20 18-56-15.703 [Debug] (Avahi) 	multicast: 1
2021-11-20 18-56-15.703 [Debug] (Avahi) 	cached: 1
2021-11-20 18-56-15.703 [Info] (Controller) Found server 192.168.16.8:1704
2021-11-20 18-56-15.703 [Info] (Connection) Resolving host IP for: 192.168.16.8
2021-11-20 18-56-15.703 [Info] (Connection) Connecting
2021-11-20 18-56-15.703 [Notice] (Connection) Connected to 192.168.16.8
2021-11-20 18-56-15.703 [Info] (Connection) My MAC: "00:1b:21:e3:a4:d6", socket: 8
2021-11-20 18-56-15.715 [Debug] (Connection) outstanding async_write
2021-11-20 18-56-15.716 [Info] (Controller) ServerSettings - buffer: 1000, latency: 0, volume: 100, muted: 0
metadata:{"STREAM":"Housewide"}
2021-11-20 18-56-15.716 [Info] (Controller) Codec: flac, sampleformat: 48000:16:2
2021-11-20 18-56-15.716 [Info] (Player) Player name: pulse, device: default, description: Let PulseAudio server choose the device, idx: 1, sharing mode: unspecified, parameters: <none>
2021-11-20 18-56-15.716 [Info] (Player) Mixer mode: software, parameters: <none>
2021-11-20 18-56-15.716 [Info] (Player) Sampleformat: 48000:16:2, stream: 48000:16:2
2021-11-20 18-56-15.716 [Info] (PulsePlayer) Setting property "application.icon_name" to "snapcast"
2021-11-20 18-56-15.716 [Info] (PulsePlayer) Setting property "media.role" to "music"
2021-11-20 18-56-15.716 [Info] (PulsePlayer) Using buffer_time: 100 ms, server: default
2021-11-20 18-56-15.717 [Debug] (PulsePlayer) State changed 2: authorizing
2021-11-20 18-56-15.722 [Debug] (PulsePlayer) State changed 3: setting name
2021-11-20 18-56-15.735 [Debug] (PulsePlayer) State changed 4: ready
2021-11-20 18-56-15.736 [Debug] (Player) setVolume exp with base 10: 1 => 1
2021-11-20 18-56-15.737 [Info] (Stream) outputBufferDacTime > bufferMs: 10555 > 1000
2021-11-20 18-56-15.737 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-15.751 [Info] (Controller) diff to server [ms]: 1.11652e+09
2021-11-20 18-56-15.752 [Info] (Stream) No chunks available
2021-11-20 18-56-16.002 [Info] (Stream) No chunks available
2021-11-20 18-56-16.754 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-17.020 [Info] (Stream) No chunks available
2021-11-20 18-56-17.771 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-18.014 [Info] (Stream) No chunks available
2021-11-20 18-56-18.787 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-19.013 [Info] (Stream) No chunks available
2021-11-20 18-56-19.804 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-20.006 [Info] (Stream) No chunks available
2021-11-20 18-56-20.817 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-21.018 [Info] (Stream) No chunks available
2021-11-20 18-56-21.822 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-22.004 [Info] (Stream) No chunks available
2021-11-20 18-56-22.827 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-23.008 [Info] (Stream) No chunks available
2021-11-20 18-56-23.836 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-24.018 [Info] (Stream) No chunks available
2021-11-20 18-56-24.846 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-25.008 [Info] (Stream) No chunks available
2021-11-20 18-56-25.861 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-26.002 [Info] (Stream) No chunks available
2021-11-20 18-56-26.875 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-27.016 [Info] (Stream) No chunks available
2021-11-20 18-56-27.886 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-28.007 [Info] (Stream) No chunks available
2021-11-20 18-56-28.903 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-56-29.003 [Info] (Stream) No chunks available
2021-11-20 18-56-31.009 [Debug] (Stats) Chunk: 0	0	0	0	14	57	0
2021-11-20 18-56-32.003 [Debug] (Stats) Chunk: 0	0	0	0	63	57	0
2021-11-20 18-56-33.018 [Debug] (Stats) Chunk: 0	0	0	0	113	57	0
2021-11-20 18-56-34.013 [Debug] (Stats) Chunk: 0	0	0	0	162	57	0
2021-11-20 18-56-35.007 [Debug] (Stats) Chunk: 0	0	0	0	211	57	0
2021-11-20 18-56-36.017 [Debug] (Stats) Chunk: 0	0	0	0	261	57	0
2021-11-20 18-56-37.006 [Debug] (Stats) Chunk: 0	0	0	0	310	57	0
2021-11-20 18-56-38.015 [Debug] (Stats) Chunk: 0	0	0	0	360	57	0
2021-11-20 18-56-39.007 [Debug] (Stats) Chunk: -1	-1	0	0	409	57	0
2021-11-20 18-56-40.014 [Debug] (Stats) Chunk: 0	0	-1	0	459	57	0
2021-11-20 18-56-41.011 [Debug] (Stats) Chunk: 0	0	0	0	500	57	0
2021-11-20 18-56-42.018 [Debug] (Stats) Chunk: -1	-1	0	0	500	57	0
2021-11-20 18-56-43.006 [Debug] (Stats) Chunk: -1	-1	-1	0	500	57	0
2021-11-20 18-56-44.004 [Debug] (Stats) Chunk: 0	0	-1	0	500	57	2
2021-11-20 18-56-45.012 [Debug] (Stats) Chunk: 0	0	0	0	500	57	2
2021-11-20 18-56-46.019 [Debug] (Stats) Chunk: 0	0	0	0	500	57	0
2021-11-20 18-56-47.011 [Debug] (Stats) Chunk: 0	0	0	0	500	57	0
2021-11-20 18-56-48.018 [Debug] (Stats) Chunk: 0	0	0	0	500	57	0
2021-11-20 18-56-49.009 [Debug] (Stats) Chunk: 0	0	0	0	500	57	0
2021-11-20 18-56-50.004 [Debug] (Stats) Chunk: 0	0	0	0	500	57	0
2021-11-20 18-56-51.015 [Debug] (Stats) Chunk: -2	-2	0	0	500	57	0
2021-11-20 18-56-52.009 [Debug] (Stats) Chunk: -3	-3	-2	0	500	57	0
2021-11-20 18-56-53.020 [Debug] (Stats) Chunk: -2	-2	-2	0	500	57	6
2021-11-20 18-56-54.009 [Debug] (Stats) Chunk: 0	0	-2	0	500	57	2
2021-11-20 18-56-55.016 [Debug] (Stats) Chunk: 0	0	0	0	500	57	0
2021-11-20 18-56-56.013 [Debug] (Stats) Chunk: 0	0	0	0	500	57	0
2021-11-20 18-56-57.011 [Debug] (Stats) Chunk: 1	1	0	0	500	57	0
2021-11-20 18-56-58.003 [Debug] (Stats) Chunk: 0	0	0	0	500	57	0
2021-11-20 18-56-59.018 [Debug] (Stats) Chunk: 0	0	0	0	500	57	0
2021-11-20 18-57-00.010 [Debug] (Stats) Chunk: 0	0	0	0	500	57	0
2021-11-20 18-57-01.002 [Debug] (Stats) Chunk: -1	-1	0	0	500	57	0
2021-11-20 18-57-02.016 [Debug] (Stats) Chunk: -1	-1	-1	0	500	57	0
2021-11-20 18-57-03.007 [Debug] (Stats) Chunk: 0	0	-1	0	500	57	1
2021-11-20 18-57-03.027 [Info] (Stream) Exception: Not enough frames available, requested frames: 965, available: 608
2021-11-20 18-57-03.027 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-57-03.047 [Debug] (Stream) age > 0: 7ms, dropping old chunks
2021-11-20 18-57-04.019 [Info] (Stream) No chunks available
2021-11-20 18-57-04.040 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-57-05.013 [Info] (Stream) No chunks available
2021-11-20 18-57-05.053 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-57-06.006 [Info] (Stream) No chunks available
2021-11-20 18-57-06.066 [Debug] (Stream) Failed to get chunk, returning silence
2021-11-20 18-57-07.019 [Info] (Stream) No chunks available
2021-11-20 18-57-07.082 [Debug] (Stream) Failed to get chunk, returning silence
^C2021-11-20 18-57-07.299 [Info] (Snapclient) Received signal 2: Interrupt
2021-11-20 18-57-07.299 [Debug] (PulsePlayer) Destructor
2021-11-20 18-57-07.299 [Info] (PulsePlayer) Stop
2021-11-20 18-57-07.299 [Debug] (PulsePlayer) State changed 6: terminated
2021-11-20 18-57-07.299 [Debug] (Connection) Disconnecting
2021-11-20 18-57-07.299 [Debug] (Connection) Disconnected
2021-11-20 18-57-07.299 [Notice] (Snapclient) Snapclient terminated.
```

**Comment by nickalcock** [comment]
Well, that's interesting. I finally understood the web interface's user interface enough to realise that you could switch between playing things by clicking on the labels (which look nothing like links, thanks, modern web design), and... my librespot stream is flawless.

So the client and server must be blameless: this must somehow be MPD's fault. This is 0.23.4, configured precisely as in the documentation:

audio_output {
        type            "fifo"
        name            "Housewide sound"
        path            "/tmp/snapfifo"
        format          "44100:16:2"
        mixer_type      "software"
}

But... crackle city.

Further analysis shows that the cause is "codec = flac" in snapserver.conf. Switch to PCM, the problem goes away. Opus is very slightly crackly.

Further analysis underway.

**Comment by nickalcock** [comment]
Presumably this has something to do with sample format mismatches -- this is, IMHO, one reason why raw PCM is a terrible format to be the only one snapcast accepts on its pipe input. It should accept *something* self-describing enough that it's got a header saying what sample format it is, so snapcast doesn't need configuring to get it right. Almost any format but PCM would do.

The annoying thing is, more or less all my music is in Vorbis format (yes, I know, I should move to Opus): MPD insists on emitting 32-bits-per-sample PCM for Vorbis, even though it's all CD-sourced so there's no way anything like that much data is actually there, and even though libvorbis is quite happy to emit 16-bit or even 8-bit PCM. (I'd pass it all the way down unchanged, but the libflac encoder can't encode 32-bits-per-sample content! AUGH). I don't think stripping the excess bits off is anything like as computationally expensive as resampling, but it's still another annoying step from bit-perfection.

I still have no idea what's up with the flac encoding. This is with libflac commit 2e7931c27eb15e387da440a37f12437e35b22dd4. What's new in libflac since then... nothing looking terribly relevant, but I'll upgrade to latest libflac master tomorrow anyway, just in case.

**Comment by nickalcock** [comment]
Really this whole bug is silly. The architecture is wrong. The player (librespot, mpd, etc) is almost always reading in compressed audio in one of a small number of well-understood formats (ffmpeg can likely decode all of them). All of those formats have timecode stamps. Rather than requiring the input to decode and likely resample the audio into a fixed sample rate, only for snapcast to recompress it and send it to a destination that then decompresses it again, the audio should be left unchanged, remuxed with data left absolutely unchanged and unresampled into a new container that supports timecodes (something libavformat can do with audio streams with minimal CPU overhead), then remuxed into a consistent format that carries timecodes and sent to the destinations, which can use those timecodes to play the audio in the same time-matched fashion they do it now.

I wonder how much work that would be... hmm... looks like ffmpeg can do an astonishing amount of it on its own. fftools/ffplay.c has an example of playback with every frame given the choice about precisely when to play it, which is where we could sync to the master clock -- though if we stored the master clock in the input pts, we'd only need to decide whether to pause in there: handling drift etc would be done by the server assigning an appropriate pts value.

Anyway, this maundering is probably out of scope for this bug :)

**Comment by nickalcock** [comment]
... I just restarted snapserver with codec = flac:1 to see if it was just taking too long to compress. No crackling. Restarted with flac:2. No crackling. Restarted with plain flac. No crackling!

I have no *idea* what is going on, but for now the crackling appears to have vanished as mysteriously as it appeared, so I might as well close this bug. If it reappears, I'll reopen.
